ulog.cpp

```diff
@@ -0,0 +1,280 @@
+    cmd.target_component = MAV_COMP_ID_ALL;
+    cmd.target_system = TARGET_SYSTEM_ID;
+
+    mavlink_msg_command_long_encode(SYSTEM_ID, 0, &msg, &cmd);
+    _endpoint->write_msg(&msg);
+
+    fflush(_file);
+    fclose(_file);
+    _file = NULL;
+    _expected_seq = 0;
+    _mainloop->set_on_master_msg_callback(NULL, NULL);
+}
+
+bool ULog::handle_fligh_stack_message(struct buffer *buffer)
+{
+    const bool mavlink2 = buffer->data[0] == MAVLINK_STX;
+    uint32_t msg_id;
+
+    if (mavlink2) {
+        struct mavlink_router_mavlink2_header *msg
+            = (struct mavlink_router_mavlink2_header *)buffer->data;
+        msg_id = msg->msgid;
+    } else {
+        struct mavlink_router_mavlink1_header *msg
+            = (struct mavlink_router_mavlink1_header *)buffer->data;
+        msg_id = msg->msgid;
+    }
+
+    /* Check if we are interested in this msg_id */
+    if (msg_id != MAVLINK_MSG_ID_COMMAND_ACK && msg_id != MAVLINK_MSG_ID_LOGGING_DATA_ACKED
+        && msg_id != MAVLINK_MSG_ID_LOGGING_DATA) {
+        return false;
+    }
+
+    /*
+     * As MAVLINK_MSG_ID_LOGGING_DATA_ACKED and MAVLINK_MSG_ID_LOGGING_DATA
+     * are MAVLink 2 only messages lets untrim all messages that match
+     * with the interested msg_ids.
+     */
+    struct buffer msg_buffer;
+    uint8_t data[MAVLINK_MAX_PACKET_LEN * 4];
+    uint8_t *payload;
+    msg_buffer.data = data;
+
+    _endpoint->untrim_msg(buffer, &msg_buffer);
+    buffer = &msg_buffer;
+
+    if (mavlink2)
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink2_header);
+    else
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink1_header);
+
+    /* Handle messages */
+    switch (msg_id) {
+    case MAVLINK_MSG_ID_COMMAND_ACK: {
+        mavlink_command_ack_t *cmd = (mavlink_command_ack_t *)payload;
+
+        if (!_timeout || cmd->command != MAV_CMD_LOGGING_START)
+            return false;
+
+        if (cmd->result == MAV_RESULT_ACCEPTED) {
+            _mainloop->timeout_del(_timeout);
+            _timeout = NULL;
+        }
+        break;
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA_ACKED: {
+        mavlink_logging_data_acked_t *ulog_data_acked = (mavlink_logging_data_acked_t *)payload;
+        mavlink_message_t msg;
+        mavlink_logging_ack_t ack;
+
+        ack.sequence = ulog_data_acked->sequence;
+        ack.target_component = MAV_COMP_ID_ALL;
+        ack.target_system = TARGET_SYSTEM_ID;
+        mavlink_msg_logging_ack_encode(SYSTEM_ID, 0, &msg, &ack);
+        _endpoint->write_msg(&msg);
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA: {
+        mavlink_logging_data_t *ulog_data = (mavlink_logging_data_t *)payload;
+        _logging_data_process(ulog_data);
+        break;
+    }
+    }
+
+    return true;
+}
+
+void ULog::_logging_data_process(mavlink_logging_data_t *msg)
+{
+    bool drops = false;
+
+    /* Check for message drops */
+    if (_expected_seq != msg->sequence) {
+        drops = true;
```

Comment:
No this needs more logic. In case of a duplicate, you can just ignore the message, since you know you already handled it before. A duplicate could be further back than just the previous message. Something like this: https://github.com/PX4/Firmware/blob/master/Tools/mavlink_ulog_streaming.py#L150 (looks somewhat ugly in python, you can probably write this better in C++).